cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramLifecycleService.java

```diff
@@ -615,8 +619,15 @@ public ProgramController start(ProgramId programId, Map<String, String> override
     return startInternal(programDescriptor, options, programRunId);
   }
 
-  private void checkCapability(ProgramDescriptor programDescriptor) throws IOException,
-    CapabilityNotAvailableException {
+  private void checkCapability(ProgramDescriptor programDescriptor) throws Exception {
+    //check for capability at application class level
+    Set<ApplicationClass> applicationClasses = artifactRepository
+      .getArtifact(Id.Artifact.fromEntityId(programDescriptor.getArtifactId())).getMeta().getClasses()
+      .getApps();
+    for (ApplicationClass applicationClass : applicationClasses) {
```

Comment:
Why requires all apps in the artifact to support the capability? Shouldn't it just be the app of the ProgramId?

Reply:
Based on offline discussion, leaving it as is because we do not have a way to associate program with a single application class. Most existing code just takes the first from the list during deployment. This should not have any consequences as of now because all application classes tagged with capability requirement has only a single Application class.